src/service/folders.rs

```diff
@@ -13,186 +13,162 @@
+pub enum FolderError {
+    /// An error that occurs while interacting with the database through the
+    /// [infra] crate.
+    #[error("InfraError# {0}")]
+    InfraError(#[from] infra::errors::Error),
+
+    /// An error that occurs when trying to set a folder name to the empty string.
+    #[error("Folder name cannot be empty")]
+    MissingName,
+
+    /// An error that occurs when trying to update the special "default" folder.
+    #[error("Can't update default folder")]
+    UpdateDefaultFolder,
+
+    /// An error that occurs when trying to delete a folder that contains dashboards.
+    #[error("Folder contains dashboards. Please move/delete dashboards from folder.")]
+    DeleteWithDashboards,
+
+    /// An error that occurs when trying to delete a folder that cannot be found.
+    #[error("Folder not found")]
+    NotFound,
+
+    /// An error occured trying to get the list of permitted folders in
+    /// enterprise mode because no user_id was provided.
+    #[error("user_id required to get permitted folders in enterprise mode")]
+    PermittedFoldersMissingUser,
+
+    /// An error occured trying to get the list of permitted folders in
+    /// enterprise mode using the validator.
+    #[error("PermittedFoldersValidator# {0}")]
+    PermittedFoldersValidator(String),
+}
+
 #[tracing::instrument(skip(folder))]
 pub async fn save_folder(
     org_id: &str,
     mut folder: Folder,
     is_internal: bool,
-) -> Result<HttpResponse, Error> {
+) -> Result<Folder, FolderError> {
     folder.name = folder.name.trim().to_string();
     if folder.name.is_empty() {
-        return Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::BAD_REQUEST.into(),
-                "folder name not allow empty".to_string(),
-            )),
-        );
+        return Err(FolderError::MissingName);
     }
 
     if !is_internal && folder.folder_id == DEFAULT_FOLDER {
-        return Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::BAD_REQUEST.into(),
-                "can't update default Dashboard folder".to_string(),
-            )),
-        );
+        return Err(FolderError::UpdateDefaultFolder);
     }
+
     if folder.folder_id != DEFAULT_FOLDER {
         folder.folder_id = ider::generate();
     }
 
-    match db::folders::put(org_id, folder).await {
-        Ok(folder) => {
-            set_ownership(org_id, "folders", Authz::new(&folder.folder_id)).await;
-            Ok(HttpResponse::Ok().json(folder))
-        }
-        Err(error) => Ok(
-            HttpResponse::InternalServerError().json(MetaHttpResponse::message(
-                http::StatusCode::INTERNAL_SERVER_ERROR.into(),
-                error.to_string(),
-            )),
-        ),
-    }
+    let folder = table::folders::put(org_id, folder).await?;
+    set_ownership(org_id, "folders", Authz::new(&folder.folder_id)).await;
+    Ok(folder)
```

Comment:
I think this is a good suggestion but for the moment I'm trying to preserve the original logic as closely as possible. 

I definitely think this would be good to do in the future, but I believe that we would want to perform the `set_ownership` check along with the call to `table::folders::put` inside a DB transaction so that the transaction (including the folder creation) could be rolled back if the call to `set_ownership` failed.